src/server/sched_func.c

```diff
@@ -0,0 +1,343 @@
+	}
+
+	return (psched);
+}
+
+/**
+ * @brief find a scheduler
+ *
+ * @param[in]	sched_name - scheduler name
+ *
+ * @return	pbs_sched *
+ */
+
+pbs_sched * find_scheduler(char *sched_name) {
+	char *pc;
+	pbs_sched *psched = NULL;
+	if (!sched_name)
+		return NULL;
+	psched = (pbs_sched *) GET_NEXT(svr_allscheds);
+	while (psched != (pbs_sched *) 0) {
+		if (strcmp(sched_name, psched->sc_name) == 0)
+			break;
+		psched = (pbs_sched *) GET_NEXT(psched->sc_link);
+	}
+	return (psched);
+}
+
+/**
+ * @brief free sched structure
+ *
+ * @param[in]	psched	- The pointer to the sched to free
+ *
+ */
+void sched_free(pbs_sched *psched) {
+	int i;
+	attribute *pattr;
+	attribute_def *pdef;
+
+	/* remove any malloc working attribute space */
+
+	for (i = 0; i < (int) SCHED_ATR_LAST; i++) {
+		pdef = &sched_attr_def[i];
+		pattr = &psched->sch_attr[i];
+
+		pdef->at_free(pattr);
+	}
+
+	/* now free the main structure */
+	delete_link(&psched->sc_link);
+	(void) free((char*) psched);
+}
+
+/**
+ * @brief - purge scheduler from system
+ *
+ * @param[in]	psched	- The pointer to the delete to delete
+ *
+ * @return	error code
+ * @retval	0	- scheduler purged
+ * @retval	PBSE_OBJBUSY	- scheduler deletion not allowed
+ */
+int sched_delete(pbs_sched *psched) {
+	pbs_db_obj_info_t obj;
+	pbs_db_sched_info_t dbsched;
+	pbs_db_conn_t *conn = (pbs_db_conn_t *) svr_db_conn;
+
+	if (psched == NULL)
+		return (0);
+
+	//TODO check for scheduler activity and return PBSE_OBJBUSY
+	/* delete scheduler from database */
+	strcpy(dbsched.sched_name, psched->sc_name);
+	obj.pbs_db_obj_type = PBS_DB_SCHED;
+	obj.pbs_db_un.pbs_db_sched = &dbsched;
+	if (pbs_db_delete_obj(conn, &obj) != 0) {
+		snprintf(log_buffer, LOG_BUF_SIZE,
+				"delete of scheduler %s from datastore failed",
+				psched->sc_name);
+		log_err(errno, __func__, log_buffer);
+	}
+	sched_free(psched);
+
+	return (0);
+}
+
+/**
+ * @brief
+ * 		action routine for the sched's "sched_port" attribute
+ *
+ * @param[in]	pattr	-	attribute being set
+ * @param[in]	pobj	-	Object on which attribute is being set
+ * @param[in]	actmode	-	the mode of setting, recovery or just alter
+ *
+ * @return	error code
+ * @retval	PBSE_NONE	-	Success
+ * @retval	!PBSE_NONE	-	Failure
+ *
+ */
+int action_sched_port(attribute *pattr, void *pobj, int actmode) {
+	if (actmode == ATR_ACTION_ALTER || actmode == ATR_ACTION_RECOV) {
+		/*TODO*/
+	}
+	return PBSE_NONE;
+}
+
+/**
+ * @brief
+ * 		action routine for the sched's "sched_priv" attribute
+ *
+ * @param[in]	pattr	-	attribute being set
+ * @param[in]	pobj	-	Object on which attribute is being set
+ * @param[in]	actmode	-	the mode of setting, recovery or just alter
+ *
+ * @return	error code
+ * @retval	PBSE_NONE	-	Success
+ * @retval	!PBSE_NONE	-	Failure
+ *
+ */
+int action_sched_priv(attribute *pattr, void *pobj, int actmode) {
+	pbs_sched* psched;
+	if (actmode == ATR_ACTION_NEW || actmode == ATR_ACTION_ALTER
+			|| actmode == ATR_ACTION_RECOV) {
```

Comment:
I just looked this up, while the action functions are called upon recovery, the return code is not used.  If your action function is only doing data validation, don't include ATR_ACTION_RECOV.  You'll just be doing extra work which will be ignored.